services/appflowy-collaborate/src/ws2/actors/workspace.rs

```diff
@@ -0,0 +1,446 @@
+impl SessionHandle {
+  fn new(collab_origin: CollabOrigin, conn: Addr<WsSession>) -> Self {
+    Self {
+      collab_origin,
+      conn,
+    }
+  }
+}
+
+pub struct Workspace {
+  server: Recipient<Terminate>,
+  workspace_id: WorkspaceId,
+  last_message_id: Rid,
+  store: Arc<CollabStore>,
+  sessions_by_client_id: HashMap<ClientID, SessionHandle>,
+  updates_handle: Option<SpawnHandle>,
+  awareness_handle: Option<SpawnHandle>,
+  snapshot_handle: Option<SpawnHandle>,
+}
+
+impl Workspace {
+  pub const SNAPSHOT_INTERVAL: std::time::Duration = std::time::Duration::from_secs(60);
+
+  pub fn new(
+    server: Recipient<Terminate>,
+    workspace_id: WorkspaceId,
+    store: Arc<CollabStore>,
+  ) -> Self {
+    Self {
+      server,
+      workspace_id,
+      store,
+      last_message_id: Rid::default(),
+      sessions_by_client_id: HashMap::new(),
+      updates_handle: None,
+      awareness_handle: None,
+      snapshot_handle: None,
+    }
+  }
+
+  async fn hande_ws_input(store: Arc<CollabStore>, sender: SessionHandle, msg: WsInput) {
+    match msg.message {
+      InputMessage::Manifest(collab_type, rid, state_vector) => {
+        match store
+          .get_latest_state(
+            msg.workspace_id,
+            msg.object_id,
+            collab_type,
+            sender.collab_origin.client_user_id().unwrap(),
+            &state_vector,
+          )
+          .await
+        {
+          Ok(state) => {
+            tracing::trace!(
+              "replying to {} manifest (client msg id: {}, server msg id: {})",
+              msg.object_id,
+              rid,
+              state.rid
+            );
+            sender.conn.do_send(WsOutput {
+              message: ServerMessage::Update {
+                object_id: msg.object_id,
+                collab_type,
+                flags: state.flags,
+                last_message_id: state.rid,
+                update: state.update,
+              },
+            });
+            tracing::trace!("sending manifest for {}", msg.object_id);
+            sender.conn.do_send(WsOutput {
+              message: ServerMessage::Manifest {
+                object_id: msg.object_id,
+                collab_type,
+                last_message_id: Rid::default(),
+                state_vector: state.state_vector,
+              },
+            });
+
+            //TODO: fetch all documents that have been updated since `rid` and send their manifests to the client
+          },
+          Err(AppError::RecordNotFound(_)) => {
+            tracing::trace!("sending manifest for new collab {}", msg.object_id);
+            sender.conn.do_send(WsOutput {
+              message: ServerMessage::Manifest {
+                object_id: msg.object_id,
+                collab_type,
+                last_message_id: Rid::default(),
+                state_vector: StateVector::default().encode_v1(),
+              },
+            });
+          },
+          Err(err) => {
+            tracing::error!(
```

Comment:
Do we need to inform the client with specific error?

Reply:
But what can client do about it?
1. We cannot close the connection, since it's shared between multiple collabs and this error is collab specific.
2. Error can be caused by temporary issues on redis/postgres/s3 and potentially be fixed in the future - but we can already do that when we open a collab or detect missing updates.

There's not much client can do about the server error here.